lib/FusionInventory/Agent/Task/Inventory/Linux/Storages.pm

```diff
@@ -115,6 +129,31 @@ sub _getDevices {
     return @devices;
 }
 
+# get serial & firmware numbers from hdparm, if available
+sub _getHdparmInfo {
+    my (%params) = @_;
+
+    return unless _correctHdparmAvailable(%params);
+
+    my %map = (
+        'serial'    => 'SERIALNUMBER',
+        'firmware'  => 'FIRMWARE',
+        'size'      => 'DISKSIZE',
+        'transport' => 'DESCRIPTION',
+        'model'     => 'MODEL',
+        'wwn'       => 'WWN'
+    );
+
+    my $hdparm = getHdparmInfo(
+        device => $params{device},
+        %params
+    );
+
+    return { map  { $map{$_} => $hdparm->{$_} }
+             grep { defined $hdparm->{$_} && exists $map{$_} }
+             keys %$hdparm };
```

Comment:
Hard to read and understand, can I suggest you to invert keys & values into %map to do:
```suggestion
    my %map = qw(
        SERIALNUMBER    serial
        FIRMWARE        firmware
        DISKSIZE        size
        DESCRIPTION     transport
        MODEL           model
        WWN             wwn
    );

    my $hdparm = getHdparmInfo(
        device => $params{device},
        %params
    )
        or return;

    foreach my $key (keys(%map)) {
        my $field = $map{$key};
        if (defined($hdparm->{$field})) {
            $map{$key} = $hdparm->{$field}
        } else {
            delete $map{$key};
        }
    }

    return \%map;
```
I think also this would be a little more efficient.

Reply:
I've come with another solution, take a look at the latest code.

Reply:
Indeed, imagine getHdparmInfo() returns a 10000 entries hash... okay this is not the case, but your %map tells only 6 entries are desired. Your code could return more entries than really expected. It works but I think it is not really optimized. Also your code will use more memory, agree this is light and just to remap fields, but...

Anyway can I propose another solution maybe far more elegant ? Just modify getHdparmInfo() in Tools/Generic.pm to report fields with expected keys ? You'll just have to also update the Win32 Storages inventory where this function is also used. Then you'll don't have to change the mapping and after all this is still what does getInfoFromSmartctl() in Tools/Linux.pm

Reply:
Done.